toolchain/check/pattern_match.cpp

```diff
@@ -63,89 +63,47 @@ class MatchContext {
+      return SemIR::InstBlockId::Empty;
+    }
+    auto block_id = context.inst_blocks().Add(results_);
+    results_.clear();
+    return block_id;
+  }
 
+  // The return slot inst emitted by `DoWork`, if any.
+  // TODO: can this be added to the block returned by `DoWork`, instead?
+  auto return_slot_id() const -> SemIR::InstId { return return_slot_id_; }
+
+ private:
   // Allocates the next unallocated RuntimeParamIndex, starting from 0.
   auto NextRuntimeIndex() -> SemIR::RuntimeParamIndex {
     auto result = next_index_;
     ++next_index_.index;
     return result;
   }
 
-  // TODO: Eliminate the caller/callee API split below, by restructuring
-  // CallerPatternMatch to operate on the whole pattern.
-
-  // Sets the result of this pattern matching operation. Must not be called when
-  // there is still pending work, except to report an error, or called more than
-  // once between calls to ConsumeResult. Valid only during caller matching.
-  auto Finish(SemIR::InstId result) -> void {
-    CARBON_CHECK(!HasWork() || result == SemIR::InstId::BuiltinError);
-    CARBON_CHECK(kind_ == MatchKind::Caller);
-    CARBON_CHECK(result_ == SemIR::InstId::Invalid);
-    result_ = result;
-  }
-
-  // Consumes and returns the result stored by Finish. Valid only during caller
-  // matching.
-  auto ConsumeResult() -> SemIR::InstId {
-    CARBON_CHECK(stack_.empty() || result_ == SemIR::InstId::BuiltinError);
-    CARBON_CHECK(kind_ == MatchKind::Caller);
-    return std::exchange(result_, SemIR::InstId::Invalid);
-  }
-
-  // Records that `bind_name_id` is the ID of an inst in the AnyBindName
-  // category, emitted as part of this pattern match. Valid only during callee
-  // pattern matching.
-  auto RecordBindName(SemIR::InstId bind_name_id) {
-    CARBON_CHECK(kind_ == MatchKind::Callee);
-    bind_name_ids_.push_back(bind_name_id);
-  }
-
-  // Allocates an InstBlock containing the IDs recorded by RecordBindName since
-  // the last call to this function (if any), and returns its ID. Valid only
-  // during callee pattern matching.
-  auto ConsumeBindNames(Context& context) -> SemIR::InstBlockId {
-    CARBON_CHECK(stack_.empty());
-    CARBON_CHECK(kind_ == MatchKind::Callee);
-    auto block_id = context.inst_blocks().Add(bind_name_ids_);
-    bind_name_ids_.clear();
-    return block_id;
-  }
-
-  auto kind() const -> MatchKind { return kind_; }
-
-  auto callee_specific_id() const -> SemIR::SpecificId {
-    return callee_specific_id_;
-  }
-
-  auto return_slot_id() const -> SemIR::InstId { return return_slot_id_; }
-
-  auto set_return_slot_id(SemIR::InstId return_slot_id) {
-    return_slot_id_ = return_slot_id;
-  }
+  auto EmitPatternMatch(Context& context, MatchContext::WorkItem entry) -> void;
 
- private:
   llvm::SmallVector<WorkItem> stack_;
 
   SemIR::RuntimeParamIndex next_index_;
 
-  SemIR::InstId result_;
-
-  llvm::SmallVector<SemIR::InstId> bind_name_ids_;
+  llvm::SmallVector<SemIR::InstId> results_;
```

Comment:
None of the private members have any comments. Would it be reasonable to add some, in order to explain the semantics? What is a "result" here?

Reply:
Good point, I've been meaning to do that. How's this?